### PR TITLE
respect DB restrictions on number of arguments in statements and queries

### PR DESCRIPTION
**Pull request by blizzz**
Follow up of https://github.com/nextcloud/server/pull/25020

Additionally to what was stated before:
- in the relevant LDAP issue run the query before there are too many parameters
- the query builder logs when there are too many parameters in total per statement (by array type parameter)

### CODE REVIEW

**Comment by blizzz** [comment]
might lead to a small conflict with https://github.com/nextcloud/server/pull/24948 (execute method in our QueryBuilder), but nothing dramatic.

File: lib/private/DB/QueryBuilder/QueryBuilder.php [hunk 1]
```diff
@@ -246,6 +246,36 @@ public function execute() {
 			}
 		}
 
+		$numberOfParameters = 0;
+		$hasTooLargeArrayParameter = false;
+		foreach ($this->getParameters() as $parameter) {
+			if(is_array($parameter)) {
```
Comment (nickvergessen) [inline lib/private/DB/QueryBuilder/QueryBuilder.php:252]:
```suggestion
			if (is_array($parameter)) {
```
making php cs happy

File: apps/user_ldap/lib/Mapping/AbstractMapping.php [hunk 2]
```diff
@@ -189,19 +190,59 @@ public function getNameByDN($fdn) {
...
+		$results = [];
+		while ($entry = $stmt->fetch(\Doctrine\DBAL\FetchMode::ASSOCIATIVE)) {
 			$results[$entry['ldap_dn']] = $entry['owncloud_name'];
 			$this->cache[$entry['ldap_dn']] = $entry['owncloud_name'];
 		}
+		$stmt->closeCursor();
+	}
+
+	public function getListOfIdsByDn(array $fdns): array {
+		$totalDBParamLimit = '65000'; //
```
Comment (nickvergessen) [inline apps/user_ldap/lib/Mapping/AbstractMapping.php:212]: You still stuttering for a comment?

Also int vs. string?

Comment (blizzz) [reply apps/user_ldap/lib/Mapping/AbstractMapping.php:212]: mh, i decided against a comment. But it should be int of course, thanks.

**Comment by blizzz** [comment]
rebased and adjusted to the dbal bump.

**Comment by MorrisJobke** [comment]
Postgres says "no".

**Comment by blizzz** [comment]
> Postgres says "no".

aye. seems to be a form of off-by-1

**Comment by blizzz** [comment]
> Postgres says "no".

Not anymore 🧑‍🔧

**Review by nickvergessen — APPROVED** [review]

File: apps/user_ldap/lib/Mapping/AbstractMapping.php [hunk 3]
```diff
@@ -199,19 +200,57 @@ public function getNameByDN($fdn) {
...
+		$fdnsSlice = count($fdns) > $sliceSize ? array_slice($fdns, 0, $sliceSize) : $fdns;
+		$qb = $this->prepareListOfIdsQuery($fdnsSlice);
+
+		while (isset($fdnsSlice[999])) {
+			// Oracle does not allow more than 1000 values in the IN list,
+			// but allows slicing
+			$slice++;
+			$fdnsSlice = array_slice($fdns, $sliceSize * ($slice - 1), $sliceSize);
+
+			if (!isset($qb)) {
```
Comment (MorrisJobke) [inline apps/user_ldap/lib/Mapping/AbstractMapping.php:236]: Psalm:

```
TypeDoesNotContainType: Cannot resolve types for $qb with type OCP\DB\QueryBuilder\IQueryBuilder and !isset assertion
```

Comment (blizzz) [reply apps/user_ldap/lib/Mapping/AbstractMapping.php:236]: It occurs to me that Psalm does not understand the `unset()`, or am I wrong?

Comment (blizzz) [reply apps/user_ldap/lib/Mapping/AbstractMapping.php:236]: Minimal example: https://psalm.dev/r/cd3ebd7a1c

**Review by rullzer — APPROVED** [review]

**Review by ChristophWurst — APPROVED** [review]
Has tests, so what could go wrong :+1: 

FYI https://www.php.net/manual/en/function.array-chunk.php seems a tad more elegant than manual slicing, but the result is the same

**Comment by blizzz** [comment]
> Has tests, so what could go wrong +1
> 
> FYI https://www.php.net/manual/en/function.array-chunk.php seems a tad more elegant than manual slicing, but the result is the same

would have a bigger memory footprint as it would return all slices at once.

**Pull request merged**